x/gov/endblocker.go

```diff
@@ -7,100 +7,87 @@ import (
 	"github.com/cosmos/cosmos-sdk/x/gov/tags"
 )
 
-// Called every block, process inflation, update validator set
+// EndBlocker called every block, process inflation, update validator set
 func EndBlocker(ctx sdk.Context, keeper Keeper) sdk.Tags {
 	logger := keeper.Logger(ctx)
 	resTags := sdk.NewTags()
 
+	// delete inactive proposal from store and its deposits
 	inactiveIterator := keeper.InactiveProposalQueueIterator(ctx, ctx.BlockHeader().Time)
-	defer inactiveIterator.Close()
-	for ; inactiveIterator.Valid(); inactiveIterator.Next() {
-		var proposalID uint64
-
-		keeper.cdc.MustUnmarshalBinaryLengthPrefixed(inactiveIterator.Value(), &proposalID)
-		inactiveProposal, ok := keeper.GetProposal(ctx, proposalID)
-		if !ok {
-			panic(fmt.Sprintf("proposal %d does not exist", proposalID))
-		}
-
-		keeper.DeleteProposal(ctx, proposalID)
-		keeper.DeleteDeposits(ctx, proposalID) // delete any associated deposits (burned)
+	keeper.IterateInactiveProposalsQueue(ctx, inactiveIterator, func(proposal Proposal) bool {
+		keeper.DeleteProposal(ctx, proposal.ProposalID)
+		keeper.DeleteProposalDeposits(ctx, proposal.ProposalID)
 
-		resTags = resTags.AppendTag(tags.ProposalID, fmt.Sprintf("%d", proposalID))
+		resTags = resTags.AppendTag(tags.ProposalID, fmt.Sprintf("%d", proposal.ProposalID))
 		resTags = resTags.AppendTag(tags.ProposalResult, tags.ActionProposalDropped)
 
 		logger.Info(
 			fmt.Sprintf("proposal %d (%s) didn't meet minimum deposit of %s (had only %s); deleted",
-				inactiveProposal.ProposalID,
-				inactiveProposal.GetTitle(),
+				proposal.ProposalID,
+				proposal.GetTitle(),
 				keeper.GetDepositParams(ctx).MinDeposit,
-				inactiveProposal.TotalDeposit,
+				proposal.TotalDeposit,
 			),
 		)
-	}
+		return false
+	})
 
 	// fetch active proposals whose voting periods have ended (are passed the block time)
 	activeIterator := keeper.ActiveProposalQueueIterator(ctx, ctx.BlockHeader().Time)
-	defer activeIterator.Close()
-	for ; activeIterator.Valid(); activeIterator.Next() {
-		var proposalID uint64
-
-		keeper.cdc.MustUnmarshalBinaryLengthPrefixed(activeIterator.Value(), &proposalID)
-		activeProposal, ok := keeper.GetProposal(ctx, proposalID)
-		if !ok {
-			panic(fmt.Sprintf("proposal %d does not exist", proposalID))
-		}
-		passes, burnDeposits, tallyResults := tally(ctx, keeper, activeProposal)
-
+	keeper.IterateActiveProposalsQueue(ctx, activeIterator, func(proposal Proposal) bool {
```

Comment:
Hmmmm, maybe this is correct, but is this the correct pattern? That is, to pass an iterator to an iterate call?

Generally, the pattern should be `x.IterateFoo(args, cb func(...){})`.

In this case, would it not make more sense to do:

```go
keeper.IterateActiveProposalsQueue(ctx, func(proposal Proposal) bool {
  // ...
})
```

This means, there is no need to pass `activeIterator` to this method. Simply have `IterateActiveProposalsQueue` create the iterator 👍

Reply:
updated